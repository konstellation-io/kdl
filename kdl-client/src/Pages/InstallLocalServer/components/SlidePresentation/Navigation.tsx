import { Button } from 'kwc';
import IconBack from '@material-ui/icons/ChevronLeft';
import IconNext from '@material-ui/icons/ChevronRight';
import React from 'react';
import cx from 'classnames';
import { range } from 'lodash';
import styles from './SlidePresentation.module.scss';

type Props = {
  actStep: number;
  totalSteps: number;
  goToStep: (step: number) => void;
  nextStep: () => void;
  prevStep: () => void;
};
function Navigation({
  actStep,
  totalSteps,
  goToStep,
  nextStep,
  prevStep,
}: Props) {
  return (
    <div className={styles.navigation}>
      <div className={styles.buttons}>
        <Button
          label=""
          Icon={IconBack}
          className={styles.button}
          onClick={prevStep}
          border
        />
        <Button
          label=""
          Icon={IconNext}
          className={styles.button}
          onClick={nextStep}
          border
        />
      </div>
      <div className={styles.dots}>
        {range(totalSteps).map((stepN) => (
          <div
            className={cx(styles.navDotContainer, {
              [styles.actual]: actStep === stepN,
            })}
            onClick={() => goToStep(stepN)}
            key={stepN}
          >
            <div className={styles.navDot} />
          </div>
        ))}
      </div>
    </div>
  );
}

export default Navigation;
