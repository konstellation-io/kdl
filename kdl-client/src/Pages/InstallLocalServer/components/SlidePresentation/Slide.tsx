import React, { useEffect } from 'react';

import styles from './SlidePresentation.module.scss';

const NEXT_SLIDE_TIMEOUT = 8000;

export type SlideInfo = {
  img: string;
  title: string;
  description: string;
};

interface Props extends SlideInfo {
  nextStep: () => void;
}
function Slide({ img, title, description, nextStep }: Props) {
  // Move to the next slide after X seconds
  useEffect(() => {
    const nextSlideInterval = setInterval(() => {
      nextStep();
    }, NEXT_SLIDE_TIMEOUT);
    return () => {
      clearInterval(nextSlideInterval);
    };
    // We want to execute this on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className={styles.slide}>
      <img className={styles.image} src={img} alt="Slide img" />
      <div className={styles.content}>
        <p className={styles.title}>{title}</p>
        <p className={styles.description} title={description}>
          {description}
        </p>
      </div>
    </div>
  );
}

export default Slide;
