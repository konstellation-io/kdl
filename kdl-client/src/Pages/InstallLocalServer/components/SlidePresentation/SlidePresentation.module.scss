@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';
@import 'Styles/animations';

$img-side: 22 * $grid-unit;

.container {
  position: relative;
  height: $img-side;
  overflow: hidden;
}

.next {
  .enter {
    animation: slideInRight 0.4s forwards;
  }
  .exit {
    animation: slideOutLeft 0.4s forwards;
  }
}

.prev {
  .enter {
    animation: slideInLeft 0.4s forwards;
  }
  .exit {
    animation: slideOutRight 0.4s forwards;
  }
}

.slide {
  display: flex;
  position: absolute;
  width: 100%;

  .image {
    border: 1px solid palette(lowlight, 500);
    width: $img-side;
    height: $img-side;
    flex-shrink: 0;
    margin-right: 2 * $grid-unit;
  }
  .content {
    flex-grow: 1;

    .title {
      @include font-title;
      color: font-color(light);

      margin: 0 0 2 * $grid-unit 0;
    }
    .description {
      @include font-small;
      color: font-color(regular);

      margin: 0;
      max-height: 12 * $grid-unit;
      overflow: hidden;
      text-overflow: ellipsis;

      display: -webkit-box;
      -webkit-line-clamp: 6; /* number of lines to show */
      -webkit-box-orient: vertical;
    }
  }
}

.navigation {
  position: absolute;
  bottom: 0;
  left: $img-side + 2 * $grid-unit;
  display: flex;
  width: -webkit-fill-available;
  background-color: palette(base, 700);

  .buttons {
    display: flex;
    margin-right: 3 * $grid-unit;

    .button:first-child {
      margin-right: 0.5 * $grid-unit;
    }

    .button {
      border-color: palette(base, 400);
      width: 3 * $grid-unit;
      height: 3 * $grid-unit !important;
      padding: 0 0 0 5px;
    }
  }
  .dots {
    display: flex;
    align-items: center;

    .navDotContainer {
      cursor: pointer;
      width: 2 * $grid-unit;
      height: 3 * $grid-unit;
      position: relative;

      &:hover .navDot {
        background-color: palette(lowlight, 600);
      }

      &.actual {
        pointer-events: none;

        .navDot {
          background-color: palette(highlight);
        }
      }

      .navDot {
        @include center;

        @include circle(palette(base, 700), 1.2 * $grid-unit);
        border: 1px solid palette(base, 200);

        transition: background-color ease 0.4s;
      }
    }
  }
}
