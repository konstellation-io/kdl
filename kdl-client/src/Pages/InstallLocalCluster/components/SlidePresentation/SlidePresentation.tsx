import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Slide, { SlideInfo } from './Slide';

import Navigation from './Navigation';
import React from 'react';
import cx from 'classnames';
import styles from './SlidePresentation.module.scss';
import useNavigation from 'Hooks/useNavigation';

type Props = {
  slides: SlideInfo[];
};
function SlidePresentation({ slides }: Props) {
  const nSlides = slides.length;

  const { actStep, goToStep, nextStep, prevStep, direction } = useNavigation({
    maxSteps: nSlides,
  });

  const slide = slides[actStep];

  return (
    <div className={styles.container}>
      <TransitionGroup className={cx(styles.pages, styles[direction])}>
        <CSSTransition
          key={actStep}
          timeout={1000}
          classNames={{
            enter: styles.enter,
            exit: styles.exit,
          }}
        >
          <Slide {...slide} nextStep={nextStep} />
        </CSSTransition>
      </TransitionGroup>
      <Navigation
        actStep={actStep}
        totalSteps={nSlides}
        goToStep={goToStep}
        nextStep={nextStep}
        prevStep={prevStep}
      />
    </div>
  );
}

export default SlidePresentation;
