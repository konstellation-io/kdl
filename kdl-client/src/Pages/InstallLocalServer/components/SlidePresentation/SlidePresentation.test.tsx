import { Button } from 'kwc';
import Navigation from './Navigation';
import React from 'react';
import Slide from './Slide';
import SlidePresentation from './SlidePresentation';
import { shallow } from 'enzyme';

const slides = [
  {
    img: 'image1',
    title: 'title1',
    description: 'description1',
  },
  {
    img: 'image2',
    title: 'title2',
    description: 'description2',
  },
];

describe('SlidePresentation component', () => {
  const component = shallow(<SlidePresentation slides={slides} />);

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show first slide at the beginning', () => {
    expect(component.find(Slide).props().title).toBe('title1');
  });

  it('handles slide updates', () => {
    component.find(Navigation).props().nextStep();
    expect(component.find(Slide).props().title).toBe('title2');
  });
});

describe('Slide component', () => {
  const component = shallow(<Slide {...slides[0]} nextStep={jest.fn()} />);

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    expect(component.contains('title1')).toBeTruthy();
    expect(component.contains('description1')).toBeTruthy();
  });
});

describe('Navigation component', () => {
  const nextStepMock = jest.fn();
  const prevStepMock = jest.fn();
  const goToStepMock = jest.fn();

  const component = shallow(
    <Navigation
      actStep={0}
      nextStep={nextStepMock}
      totalSteps={2}
      goToStep={goToStepMock}
      prevStep={prevStepMock}
    />
  );

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right components', () => {
    expect(component.find('.navDotContainer').length).toBe(2);
  });

  it('moves to between steps', () => {
    component.find(Button).last().simulate('click');
    expect(nextStepMock).toHaveBeenCalledTimes(1);

    component.find(Button).first().simulate('click');
    expect(prevStepMock).toHaveBeenCalledTimes(1);

    component.find('.navDotContainer').last().simulate('click');
    component.find('.navDotContainer').first().simulate('click');
    expect(goToStepMock).toHaveBeenCalledTimes(2);
  });
});
