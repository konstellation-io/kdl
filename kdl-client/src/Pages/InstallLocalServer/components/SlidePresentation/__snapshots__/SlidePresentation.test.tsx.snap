// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component Component match snapshot 1`] = `
<div
  className="navigation"
>
  <div
    className="buttons"
  >
    <P
      Icon={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "displayName": "ChevronLeftIcon",
          "muiName": "SvgIcon",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      border={true}
      className="button"
      label=""
      onClick={[MockFunction]}
    />
    <P
      Icon={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "displayName": "ChevronRightIcon",
          "muiName": "SvgIcon",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      border={true}
      className="button"
      label=""
      onClick={[MockFunction]}
    />
  </div>
  <div
    className="dots"
  >
    <div
      className="navDotContainer actual"
      key="0"
      onClick={[Function]}
    >
      <div
        className="navDot"
      />
    </div>
    <div
      className="navDotContainer"
      key="1"
      onClick={[Function]}
    >
      <div
        className="navDot"
      />
    </div>
  </div>
</div>
`;

exports[`Slide component Component match snapshot 1`] = `
<div
  className="slide"
>
  <img
    alt="Slide img"
    className="image"
    src="image1"
  />
  <div
    className="content"
  >
    <p
      className="title"
    >
      title1
    </p>
    <p
      className="description"
      title="description1"
    >
      description1
    </p>
  </div>
</div>
`;

exports[`SlidePresentation component Component match snapshot 1`] = `
<div
  className="container"
>
  <TransitionGroup
    childFactory={[Function]}
    className="pages next"
    component="div"
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "enter",
          "exit": "exit",
        }
      }
      key="0"
      timeout={500}
    >
      <Slide
        description="description1"
        img="image1"
        nextStep={[Function]}
        title="title1"
      />
    </CSSTransition>
  </TransitionGroup>
  <Navigation
    actStep={0}
    goToStep={[Function]}
    nextStep={[Function]}
    prevStep={[Function]}
    totalSteps={2}
  />
</div>
`;
